presto-spi/src/main/java/com/facebook/presto/spi/type/TypeSignature.java

```diff
@@ -177,7 +177,10 @@ private static TypeSignature parseRowTypeSignature(String signature, Set<String>
             else if (c == ' ') {
                 if (bracketCount == 1 && inFieldName) {
                     checkArgument(parameterStart >= 0 && parameterStart < i, "Bad type signature: '%s'", signature);
-                    fieldName = signature.substring(parameterStart, i);
+                    // The field name identifier may have quotes at this stage but they are ignored
+                    // while creating the type signature. This ensures that signatures for quoted and
+                    // unquoted versions of the same field name will match.
+                    fieldName = signature.substring(parameterStart, i).replaceAll("^\"|\"$", "");
```

Comment:
Unfortunately, while this solves the immediate problem, it's not the correct fix for how the feature should work.

Field names in row types are identifiers, which can be regular or delimited (quoted). Quoting matters because it affects how the resolution and matching rules work. In particular:

```
Row type RT2 is a subtype of data type RT1 if and only if RT1 and RT2 are row types of 
the same degree and, in every n-th pair of corresponding field definitions, FD1n in RT1 
and FD2n in RT2, the <field name>s are equivalent and the <data type> of FD2n is a 
subtype of the <data type> of FD1n.
```

And:

```
25) The case-normal form of the <identifier body> of a <regular identifier> is used for 
  purposes such as and including determination of identifier equivalence, representation 
  in the Definition and Information Schemas, and representation in diagnostics areas.

26) The case-normal form of <regular identifier> shall not be equal, according to the 
  comparison rules in Subclause 8.2, “<comparison predicate>”, to any <reserved word> 
  (with every letter that is a lower-case letter replaced by the corresponding upper-case 
  letter or letters), treated as the repetition of a <character string literal> that specifies a 
  <character set specification> of SQL_IDENTIFIER.

27) Two <regular identifier>s are equivalent if the case-normal forms of their 
  <identifier body>s, considered as the repetition of a <character string literal> that specifies 
  a <character set specification> of SQL_IDENTIFIER and an implementation-defined collation 
  IDC that is sensitive to case, compare equally according to the comparison rules in 
  Subclause 8.2, “<comparison predicate>”.

28) A <regular identifier> and a <delimited identifier> are equivalent if the case-normal form 
  of the <identifier body> of the <regular identifier> and the <delimited identifier body> of the 
  <delimited identifier> (with all occurrences of <quote> replaced by <quote symbol> and all 
  occurrences of <doublequote symbol> replaced by <double quote>), considered as the 
  repetition of a <character string literal> that specifies a <character set specification> of 
  SQL_IDENTIFIER and IDC, compare equally according to the comparison rules in Subclause 
  8.2, “<comparison predicate>”.

29) Two <delimited identifier>s are equivalent if their <delimited identifier body>s, considered 
  as the repetition of a <character string literal> that specifies a <character set specification> 
  of SQL_IDENTIFIER and an implementation-defined collation that is sensitive to case, 
  compare equally according to the comparison rules in Subclause 8.2, “<comparison predicate>”.
```

An interesting data point is that row type fields need to behave exactly the same as how columns in tables work:

```
The most specific type of a row of a table is a row type. In this case, each column of the 
table corresponds to the field of the row type that has the same ordinal position as the column.
```

Reply:
I believe the concern is mainly regarding the syntax rules 26, 28 and 29. (Please correct me if I am wrong)

Rule 26 is talking specifically about a case-normal \<regular identifier\>  and reserved words. I could not find any rule regarding reserved words and \<delimited identifier\>. If the identifier is not delimited and matches a reserved word we already see a syntax error during the parsing. Based on the usage of TypeSignature I could not easily identify any context that would be problematic but It's possible I may be missing something. 

For rule 28 & 29 the I can make changes in the PR as needed. Mainly to handle \<doublequote symbol\> appearing in \<delimited identifier body\>

Regarding the resolution rules for row types.
>Row type RT2 is a subtype of data type RT1 if and only if...

If I understand correctly the concern here is not regarding subtype resolution and matching as I don't think Presto has support for what sql defines as user defined types (which can have subtypes).  Instead the concern is we are removing quotes from delimited identifiers in the TypeSignature ?

> SQL supports three sorts of data types: predefined data types, constructed types, and user-defined types. Prede-
fined data types are sometimes called “built-in data types”, though not in this International Standard. User-
defined types can be defined by a standard, by an implementation, or by an application.
A constructed type is specified using one of SQL's data type constructors, ARRAY, MULTISET, REF, and
ROW. A constructed type is either an array type, a multiset type, a reference type, or a row type, according to
whether it is specified with ARRAY, MULTISET, REF, or ROW, respectively. Array types and multiset types
are known generically as collection types.

Please let me know your thoughts on whether this approach of  modifying the TypeSignature  would be acceptable with a few fixes (mentioned above) or if there is an alternative I should consider?